optuna/_updated_trials_queue.py

```diff
@@ -0,0 +1,115 @@
+from collections import deque
+import threading
+from typing import Any
+from typing import Deque
+from typing import Dict
+from typing import List
+from typing import Tuple
+from typing import TYPE_CHECKING
+
+
+if TYPE_CHECKING:
+    from optuna import Study
+    from optuna.trial import FrozenTrial
+    from optuna.trial import TrialState
+
+
+class UpdatedTrialsQueue(object):
+    """A virtual queue of trials in the specified states.
+
+    This class imitates a queue that trial is added when it goes into specified states.
+
+    It is not supposed to be directly accessed by library users except to write user-defined
+    samplers.
+
+    Note that the ``states`` argument should consist of the same stage states.
+    """
+
+    def __init__(self, study: "Study", states: Tuple["TrialState", ...]) -> None:
+        for state in states:
+            if state < states[0] or states[0] < state:
+                raise RuntimeError("The states should be in the same stage.")
+
+        self._study = study
+        self._states = states
+
+        self._queue: Deque[int] = deque()
+        self._watching_trial_indices: List[int] = []
+        self._next_min_trial_index = 0
+
+        self._lock = threading.Lock()
+
+    def __getstate__(self) -> Dict[Any, Any]:
+
+        state = self.__dict__.copy()
+        del state["_lock"]
+        return state
+
+    def __setstate__(self, state: Dict[Any, Any]) -> None:
+
+        self.__dict__.update(state)
+        self._lock = threading.Lock()
+
+    def _fetch_trials(self, deepcopy: bool) -> List["FrozenTrial"]:
+        trials = self._study.get_trials(deepcopy=deepcopy)
```

Comment:
Here, we're still constructing all trials. How about taking some benchmarks before going forward to understand in what practical use cases this feature still becomes beneficial?

Reply:
@hvy Yes, benchmarks are necessary. I will work on it with another PR.

@HideakiImamura I'm not sure, but it may be slow to check `self._watching_trial_indices` using `get_trial` one by one.

In specific cases, it has room to improve performance. For example, `_pop_waiting_trial_id` checks a trial is available after getting it. Therefore, if `self._queue` is not empty, we can return a trial in it without `self._fetch_trials`.